### DISCUSSION TITLE
jQuery 3, o que acharam?

### DISCUSSION THREAD

**Discussion by felipefialho** [frontendbr/forum#259] [Discussão]
[Para quem ainda não viu](http://blog.jquery.com/2016/06/09/jquery-3-0-final-released/)

Sei que muita gente torce o nariz para o jQuery, mas o fato é que continua sendo amplamente utilizado e agora tem uma nova versão cheia de novidades.

Já chegaram a dar uma olhada? O que acharam?

@Mendrone @grippado @matmarsiglio: Trazendo nossa discussão de sexta pra cá.

**Comment by grippado** [comment]
Tem uma lista de features fodas na $3, mas a que eu curti muito.. e que ajuda bastante a evitar problemas com $.ajax, é a assinatura de função $.get() e $.post, evitando "mistakes" de fazer POST e/ou GET dentro do $.ajax() de forma equivocada, e que pode causar problemas em serviços!

entre outras cossitas como: poder usar :hidden e :visible dentro dos seletores, e eles retornarem algo válido!
(lembram dos problemas de assinar um função no document(ready), pra uma tag invisível??
Com esses seletores, vc avisa que o objeto está 'hidden' e ele é assinado, evitando bastante transtorno!

Tem muitos pontos bacanas pra serem citados, esses foram alguns que eu achei relevante comentar!
=)

**Comment by renatoagds** [comment]
Eu particularmente mexi pouco nessa nova versão, pelo changelog achei bacana, principalmente por eles estarem usando Promises. 

Gostaria de ver se alguém está usando em Prod e o que melhorou se sim.

**Comment by felipefialho** [comment]
Como trabalho em agencia por vezes fazemos uso do jQuery, para projetos que o Zepto por si só não satisfaz ou não tem compatibilidade com algo que vamos aplicar.

Mas ainda estamos usando o 2. Ainda não testamos o 3 e também fiquei curioso se alguém já aplicou em produção.

**Comment by lagden** [comment]
Eu uso bastante **AMD** e uso o **jQuery** dessa maneira também!
O que eu mais gostava era o `$.Defer` que agora virou `Promise`, então totalmente excelente!!

**Zepto** is bullshit (refazendo a roda)!!

Artigo bacana pra usar o **jQuery** com **AMD**, assim vocês nunca mais usarão o zepto...

http://developer.telerik.com/featured/jquery-using-only-what-you-need/

**Comment by felipefialho** [comment]
@lagden Tivemos problemas de compatibilidade com diversas features que precisávamos quando usamos o Zepto e acabamos indo para o jQuery de novo.

Deixei o Zepto só para projetos estáticos (ou quase isso) que demorariam mais para fazer com Vanilla (tempo é R$), já que é bem leve.

EDIT: E essa ideia de usar com **AMD** é excelente e é engraçado pensar que uso o Zepto assim, mas nunca pensei em usar o jQuery, vai entender, hahaa.

**Comment by lagden** [comment]
@LFeh sim... ele foi bom na época em que não dava para gerar um jQuery customizado ou utilizar como módulo...
Vida longa ao jQuery, pois as diferenças sempre existirão!!

**Comment by mtmr0x** [comment]
boa @LFeh!

Eu sou a favor de usar o que é bom pro projeto. Olhei o change log deles por cima e achei massa também, nada surpreendente, só um _"finalmente, hein!?"_

Só que admito que vir com Promises é uma parada que por mais que esperado, tem seu mérito. Eu vi que está bacana e não vi sobre quanto ficou mais leve ou mais pesado o tamanho final do $3 para usar em prod, só que minha maior bronca com o jQuery em si é performance e pelo que vi, eles estão melhorando isso bastante.

E só pra finalizar, o maior problema do jQuery é o dev, que não sabe que quando for procurar um elemento no DOM, ele precisa fazer `$('.el').find('.child')` e não `$('.el .child')`, para questões de performance (e até boas práticas nas convenções da comunidade de jQuery).

Então parabéns ao jQuery e espero que alguém esteja usando em prod para falar suas XP's e problemas que encontraram até com codigo $2 antigo e essa parte de atualização em projetos que já usam o $2.

**Comment by vitor-mendrone-deel** [comment]
O jQuery ainda é a realidade da maioria dos Devs. 

Acho que vivemos em um ambiente onde parece que todos estão usando ES6, React, etc, o que não é real. 

Acho excelente ver o jQuery se atualizando, o lance das promises IMHO é o maior destaque

Um dos problemas principais de quando se usava o jQuery.deferred era a forma como o "bubbling" nas exceptions cancelava a execução dos callbacks e com promises isso deixa de ocorrer. 

É algo útil até mesmo pra quem está trampando com React e usando jQuery para as chamadas Ajax, já que com a promise você evita erros no mount e unmount dos components

**Comment by renatoagds** [comment]
@lagden você chegou a usar em prod? 

Vou da mesma visão que o @matmarsiglio, foi legal a atualização, mas nada muito bombastico.

**Comment by lagden** [comment]
@renatoagds o $.3? já estou usando no projeto... daqui 1 mês estará no ar!! :metal:

**Comment by olegon** [comment]
@Mendrone Concordo com a visão sobre a realidade dos Devs. Um dos projetos da onde eu trabalho tem como _target_ máquinas WinXP + IE6, então até o jQuery 1.x.x ainda é realidade. O front-end do sistema é muito simples e não há muito o que fazer.

**Comment by renatoagds** [comment]
@lagden Quando colocar ele no ar, lembra de vir aqui para nos falar como foi! :+1:

**Comment by eduardojmatos** [comment]
O que eu curti bastante:
- o lance de usar o `requestAnimationFrame` para as animações;
- remoção do `$.parseJSON` pra usar o `JSON.parse`;
- drop no suporte a browsers antigos;
- touch events (enfim!!!);

\* @grippado os aliases `$.get` e `$.post` já existiam antes da version 3 <3

**Comment by grippado** [comment]
@eduardojmatos eu sei que existiam, eu não disse que eles são novos, eu falei sobre as novas assinaturas das funções pra evitar problemas, vou tentar explicar:
As assinaturas são:

``` javascript
$.get([settings])
$.post([settings])
```

Se vc tiver o código seguinte!

``` javascript
$.get({
   url: 'https://www.teste.post',
   method: 'POST'
});
```

Apesar da propriedade de método, esta afirmação não vai fazer um POST, mas um GET, por causa das novas 'signatures'.

Espero que tenha conseguido explicar direitinho! rs

**Comment by eduardojmatos** [comment]
@grippado entendi mano! Que gambiarra que era isso hein... huahua

**Comment by olegon** [comment]
> - o lance de usar o requestAnimationFrame para as animações;

Esse item é beeeem interessante!

**Comment by vitor-mendrone-deel** [comment]
Tem uma parada complicada no jQuery 3 que é a mudança na função jQuery(document).ready

Basicamente essa função agora tem internamente uma promise deferida que esconde as exceptions que rolam no callback. 

Algo tipo: 

``` javascript
$(function() {
    throw new Error("fodeu!")
    $('elemento').text("cade o erro, cuzão?");
})
```

Vai resultar em uma tela em branco e em um console sem erros (no jQuery 2 a tela também ficaria em branco mas o console mostraria uma exception). 

Isso tem uma solução mas é verbosa pra cacete, então eu não recomendaria usar jQuery 3 pra inicialização de uma aplicação.

**Comment by vitor-mendrone-deel** [comment]
@grippado eu não tinha ideia desse lance com essa funções mano. Thanks for sharing! :heart:

**Comment by renatoagds** [comment]
@grippado não sabia desse lance também, você chegou a usar em algum lugar? Rolou algum case desse caso? Maneiro :)
